datafusion/expr/src/logical_plan/plan.rs

```diff
@@ -253,9 +253,12 @@ impl LogicalPlan {
                 aggr_expr,
                 ..
             }) => group_expr.iter().chain(aggr_expr.iter()).cloned().collect(),
+            // There are two part of expression for join, equijoin(on) and non-equijoin(filter).
+            // 1. the first part is `on.len()` equijoin expressions, and the struct of each expr is `left-on = right-on`.
+            // 2. the second part is non-equijoin(filter).
             LogicalPlan::Join(Join { on, filter, .. }) => on
                 .iter()
-                .flat_map(|(l, r)| vec![l.clone(), r.clone()])
+                .map(|(l, r)| Expr::eq(l.clone(), r.clone()))
```

Comment:
This is the fix, right? It then exposes the `<l> = <r>` expr to the existing type coercion logic ?

Very nice 👍